File: src/client/java/dev/ilgax/wynnhidepet/mixin/client/ClientPlayerInteractionManagerMixin.java
package dev.ilgax.wynnhidepet.mixin.client;

import dev.ilgax.wynnhidepet.ModConfigKt;
import dev.ilgax.wynnhidepet.client.PetEntityTracker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.MultiPlayerGameMode;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.EntityHitResult;
import net.minecraft.world.phys.Vec3;
import org.jetbrains.annotations.Nullable;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Unique;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.util.Optional;

@Mixin(MultiPlayerGameMode.class)
public class ClientPlayerInteractionManagerMixin {

    @Unique
    private long wynnhidepet_lastInteractTick = -1L;

    @Inject(method = "attack", at = @At("HEAD"), cancellable = true)
    private void onAttackEntity(Player player, Entity entity, CallbackInfo ci) {
        if (ModConfigKt.getConfig().getHidePets() && PetEntityTracker.INSTANCE.getPetEntityIds().contains(entity.getId())) {
            ci.cancel();
        }
    }

    @Inject(method = "interact", at = @At("HEAD"), cancellable = true)
    private void onInteractEntity(Player player, Entity entity, InteractionHand hand, CallbackInfoReturnable<InteractionResult> cir) {
        if (!ModConfigKt.getConfig().getHidePets() || !PetEntityTracker.INSTANCE.getPetEntityIds().contains(entity.getId())) return;
        Minecraft client = Minecraft.getInstance();
        if (client.level == null) { cir.setReturnValue(InteractionResult.PASS); return; }
        long currentTick = client.level.getGameTime();
        if (wynnhidepet_lastInteractTick == currentTick) { cir.setReturnValue(InteractionResult.PASS); return; }
        
        Entity behind = findEntityBehindPets(player);
        if (behind != null) {
            MultiPlayerGameMode im = client.gameMode;
            if (im != null) {
                wynnhidepet_lastInteractTick = currentTick;
                InteractionResult result = im.interact(player, behind, hand);
                cir.setReturnValue(result);
                return;
            }
        }
        
        // If no entity behind or no gameMode, consume the click so it doesn't reach the pet
        cir.setReturnValue(InteractionResult.PASS);
    }

    @Inject(method = "interactAt", at = @At("HEAD"), cancellable = true)
    private void onInteractEntityAtLocation(Player player, Entity entity, EntityHitResult hitResult, InteractionHand hand, CallbackInfoReturnable<InteractionResult> cir) {
        if (!ModConfigKt.getConfig().getHidePets() || !PetEntityTracker.INSTANCE.getPetEntityIds().contains(entity.getId())) return;
        Minecraft client = Minecraft.getInstance();
        if (client.level == null) { cir.setReturnValue(InteractionResult.PASS); return; }
        long currentTick = client.level.getGameTime();
        if (wynnhidepet_lastInteractTick == currentTick) { cir.setReturnValue(InteractionResult.PASS); return; }

        Entity behind = findEntityBehindPets(player);
        if (behind != null) {
            MultiPlayerGameMode im = client.gameMode;
            if (im != null) {
                wynnhidepet_lastInteractTick = currentTick;
                InteractionResult result = im.interact(player, behind, hand);
                cir.setReturnValue(result);
                return;
            }
        }

        // Always block/pass on the pet itself
        cir.setReturnValue(InteractionResult.PASS);
    }

    @Unique
    @Nullable
    private Entity findEntityBehindPets(Player player) {
        Minecraft client = Minecraft.getInstance();
        if (client.level == null) return null;

        // Mojang 1.21.1: getEntityInteractionRange() -> entityInteractionRange()
        double reach = player.entityInteractionRange();
        Vec3 eyePos = player.getEyePosition();
        Vec3 lookVec = player.getViewVector(1.0f);
        Vec3 endPos = eyePos.add(lookVec.multiply(reach, reach, reach));

        // Use a tighter bounding box for the initial search
        AABB searchBox = player.getBoundingBox().expandTowards(lookVec.scale(reach)).inflate(0.5);

        Entity closest = null;
        double closestDistSq = reach * reach;

        // Filter out pets early to avoid redundant raycasts.
        for (Entity e : client.level.getEntities(player, searchBox,
                ent -> !PetEntityTracker.INSTANCE.getPetEntityIds().contains(ent.getId()) && ent.isPickable())) {

            Optional<Vec3> hit = e.getBoundingBox().inflate(e.getPickRadius()).clip(eyePos, endPos);
            if (hit.isPresent()) {
                double distSq = eyePos.distanceToSqr(hit.get());
                if (distSq < closestDistSq) {
                    closestDistSq = distSq;
                    closest = e;
                }
            }
        }

        return closest;
    }
}

File: src/client/kotlin/dev/ilgax/wynnhidepet/mixin/client/ClientPlayerInteractionManagerMixin.kt
package dev.ilgax.wynnhidepet.mixin.client

import dev.ilgax.wynnhidepet.client.PetEntityTracker
import dev.ilgax.wynnhidepet.getConfig
import net.minecraft.client.MinecraftClient
import net.minecraft.client.network.ClientPlayerInteractionManager
import net.minecraft.entity.Entity
import net.minecraft.entity.attribute.EntityAttributes
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.EntityHitResult
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(ClientPlayerInteractionManager::class)
class ClientPlayerInteractionManagerMixin {

    @Inject(method = ["attackEntity"], at = [At("HEAD")], cancellable = true)
    fun onAttackEntity(
        player: PlayerEntity,
        entity: Entity,
        ci: CallbackInfo
    ) {
        if (getConfig().hidePets && entity.id in PetEntityTracker.petEntityIds) {
            ci.cancel()
        }
    }

    @Inject(method = ["interactEntity"], at = [At("HEAD")], cancellable = true)
    fun onInteractEntity(
        player: PlayerEntity,
        entity: Entity,
        hand: Hand,
        cir: CallbackInfoReturnable<ActionResult>
    ) {
        if (!getConfig().hidePets || entity.id !in PetEntityTracker.petEntityIds) return
        val behindEntity = findEntityBehindPets(player) ?: run { cir.returnValue = ActionResult.PASS; return }
        MinecraftClient.getInstance().interactionManager?.interactEntity(player, behindEntity, hand)
        cir.returnValue = ActionResult.SUCCESS
    }

    @Inject(method = ["interactEntityAtLocation"], at = [At("HEAD")], cancellable = true)
    fun onInteractEntityAtLocation(
        player: PlayerEntity,
        entity: Entity,
        hitResult: EntityHitResult,
        hand: Hand,
        cir: CallbackInfoReturnable<ActionResult>
    ) {
        if (!getConfig().hidePets || entity.id !in PetEntityTracker.petEntityIds) return
        val behindEntity = findEntityBehindPets(player) ?: run { cir.returnValue = ActionResult.PASS; return }
        MinecraftClient.getInstance().interactionManager?.interactEntity(player, behindEntity, hand)
        cir.returnValue = ActionResult.SUCCESS
    }

    private fun findEntityBehindPets(player: PlayerEntity): Entity? {
        val client = MinecraftClient.getInstance()
        val world = client.world ?: return null
        val reach = player.getAttributeValue(EntityAttributes.ENTITY_INTERACTION_RANGE)
        val eyePos = player.eyePos
        val lookVec = player.getRotationVec(1.0f)
        val endPos = eyePos.add(lookVec.multiply(reach))
        val searchBox = player.boundingBox.stretch(lookVec.multiply(reach)).expand(1.0, 1.0, 1.0)

        var closest: Entity? = null
        var closestDist = Double.MAX_VALUE

        for (entity in world.getOtherEntities(player, searchBox)) {
            if (entity.id in PetEntityTracker.petEntityIds) continue
            if (!entity.canHit()) continue
            val hit = entity.boundingBox.expand(0.3).raycast(eyePos, endPos)
            if (hit.isPresent) {
                val dist = eyePos.squaredDistanceTo(hit.get())
                if (dist < closestDist) {
                    closestDist = dist
                    closest = entity
                }
            }
        }

        return closest
    }
}
